spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java

```diff
@@ -820,6 +830,9 @@ protected Producer<K, V> createKafkaProducer() {
 	 * @since 2.2.13
 	 */
 	protected final boolean removeProducer(CloseSafeProducer<K, V> producerToRemove, Duration timeout) {
+		if (producerToRemove.closed) {
```

Comment:
This method is `protected`, so, can be overridden in the target project.
Therefore, `this.listeners.forEach` might not be closed.

I'm not sure why you have changed that `closeDelegate()` implementation, but probably it is not what we really need for the functionality you trying to implement.

Reply:
> This method is protected, so, can be overridden in the target project.
Therefore, this.listeners.forEach might not be closed.

I'm sorry, I didn't quite understand your point. What do you mean by "this.listeners.forEach might not be closed"?

> I'm not sure why you have changed that closeDelegate() implementation, but probably it is not what we really need for the functionality you trying to implement.

[This](https://github.com/spring-projects/spring-kafka/pull/2930#discussion_r1428924996) is the main reason why I change `closeDelegate()` implementation. The goal is to ensure that the behavior remains unchanged while providing a more user-friendly approach to closing the producer. Otherwise, I would need to duplicate the logic for the close operation to implement my feature.

Reply:
Sorry. I mean `not called`. 
Well, I see now that this `removeProducer` is `final`. So, it cannot be overridden.
I still don't understand what drove you to do that, but I probably will believe you as is.